### PR TITLE
service/dap: support clearing breakpoints and setting breakpoint conditions

### PR DESCRIPTION
**Pull request by polinasok**
Updates #1515

### CODE REVIEW

File: service/dap/server.go [hunk 1]
```diff
@@ -546,9 +582,9 @@ func (s *Server) onSetBreakpointsRequest(request *dap.SetBreakpointsRequest) {
 	i := 0
 	for _, b := range request.Arguments.Breakpoints {
 		bp, err := s.debugger.CreateBreakpoint(
-			&api.Breakpoint{File: request.Arguments.Source.Path, Line: b.Line})
+			&api.Breakpoint{File: request.Arguments.Source.Path, Line: b.Line, Cond: b.Condition})
 		if err != nil {
-			s.log.Error("ERROR:", err)
+			s.log.Error("ERROR: ", err)
```
Comment (aarzilli) [inline service/dap/server.go:587]: I took a look at the SetBreakpointsRequest specification and it seems to me that when there is an error the adapter can return `verified == false` and set the `message` field to explain why the breakpoint couldn't be verified, wouldn't that be better?

I'm approving this, this is just something to consider.

**Review by aarzilli — APPROVED** [review]

File: service/dap/server.go (hunk 1 above)
Comment (polinasok) [reply service/dap/server.go]: Done

File: service/dap/server.go [hunk 2]
```diff
@@ -535,9 +536,44 @@ func (s *Server) onDisconnectRequest(request *dap.DisconnectRequest) {
...
-		s.log.Error("ERROR: Unable to set breakpoint for empty file path")
+		s.sendErrorResponse(request.Request, UnableToSetBreakpoints, "Unable to set or clear breakpoints", "empty file path")
+		return
+	}
+
+	// According to the spec we should "set multiple breakpoints for a single source
+	// and clear all previous breakpoints in that source." The simplest way is
+	// to clear all and then set all.
+	//
+	// TODO(polina): should we optimize this as follows?
```
Comment (polinasok) [inline service/dap/server.go:550]: Do you have an opinion here? (This was a user request).

Comment (aarzilli) [reply service/dap/server.go:550]: I'm skeptical that the performance difference would be visible. Maybe when the adapter was javascript talking to delve using our API over laggy network.

**Pull request merged**